File: codeforces/1900~1999/1975/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  int a = 0, b = 0;
  cin >> a >> b;

  vector<vector<int>> adj(n+1);
  for (int i = 1; i < n; ++i) {
    int u = 0, v = 0;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
  }

  function<void(int,int,vector<int>&)> dfs = [&](int node, int p, vector<int>& d) {
    for (int next_node : adj[node]) if (next_node != p) {
      d[next_node] = d[node] + 1;
      dfs(next_node, node, d);
    }
  };

  vector<int> a_dist(n+1);
  dfs(a, -1, a_dist);
  vector<int> b_dist(n+1);
  dfs(b, -1, b_dist);

  auto find_meeting_point = [&]() -> int {
    vector<bool> visited(n+1);
    queue<int> q;
    q.push(b);
    visited[b] = true;
    int step = 0;
    while (!q.empty()) {
      for (int size = q.size(); size > 0; --size) {
        int node = q.front();
        q.pop();

        if (a_dist[node] <= step) {
          return node;
        }

        for (int next_node : adj[node]) {
          if (visited[next_node]) continue;
          visited[next_node] = true; 
          q.push(next_node);
        }
      }
      step++;
    }
    return 0;
  };

  int meet = find_meeting_point();

  vector<int> meet_dist(n+1);
  dfs(meet, -1, meet_dist);

  int farthest = *max_element(meet_dist.begin(), meet_dist.end());
  int ans = 2 * (n - 1) - farthest + b_dist[meet];
  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1900~1999/1975/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::collections::VecDeque;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn compute_dist(node: usize, parent: usize, adj: &Vec<Vec<usize>>, d: &mut Vec<i32>) {
    for &next_node in adj[node].iter() {
        if next_node == parent {
            continue;
        }

        d[next_node] = d[node] + 1;
        compute_dist(next_node, node, adj, d);
    }
}

fn find_meeting_point(b: usize, adj: &Vec<Vec<usize>>, a_dist: &Vec<i32>) -> usize {
    let n = adj.len();
    let mut q = VecDeque::new();
    q.push_back(b);
    let mut visited = vec![false; n+1];
    let mut step = 0;
    while !q.is_empty() {
        let size = q.len();
        for _i in 0..size {
            if let Some(node) = q.pop_front() {
                if a_dist[node] <= step {
                    return node;
                }

                for &next_node in adj[node].iter() {
                    if visited[next_node] {
                        continue;
                    }
                    visited[next_node] = true;
                    q.push_back(next_node);
                }
            }
        }
        step += 1;
    }
    0
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let n = read_int::<usize>();
    let inputs = read_vec::<usize>();
    let a = inputs[0];
    let b = inputs[1];

    let mut adj = vec![Vec::new(); n+1];
    for _i in 1..n {
        let edge = read_vec::<usize>();
        let u = edge[0];
        let v = edge[1];
        adj[u].push(v);
        adj[v].push(u);
    }
    
    let mut a_dist = vec![0; n+1];
    compute_dist(a, 0, &adj, &mut a_dist);
    let mut b_dist = vec![0; n+1];
    compute_dist(b, 0, &adj, &mut b_dist);

    let meet = find_meeting_point(b, &adj, &a_dist);
    let mut meet_dist = vec![0; n+1];
    compute_dist(meet, 0, &adj, &mut meet_dist);

    let d = *meet_dist.iter().max().unwrap();
    let ans = 2 * (n as i32 - 1) - d + b_dist[meet];
    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
